Add Pluto-Eris cycle of curves

davidnevadoc opened the pull request:
Implementation of the half-pairing cycle of prime-order curves described here:  https://github.com/daira/pluto-eris

Pluto is a Barreto–Naehrig curve and its pairing implementation closely resembles that of `bn256` curve. See the Optimal Ate pairing described in the article: https://eprint.iacr.org/2010/354.pdf.

The constants for the fields and its extensions are derived these notebooks: https://github.com/davidnevadoc/ec-constants/tree/main/pluto_eris

CPerezz commented on src/pluto_eris/fields/mod.rs:
```diff
@@ -0,0 +1,848 @@
+            #[inline]
+            pub const fn mul(&self, rhs: &Self) -> $field {
+                // Schoolbook multiplication
+
```
```suggestion
```

CPerezz reviewed (changes requested):
First of all, nice work. It's really cool to get these curves implemented!

I have a couple of things that have written down while reviewing:
- README.md needs to be updated mentioning that pluto/eris curves are supported.
- All tests should be inside test modules please.
- Add comments for functions left unimplemented. Why were they unimplemented?
- Add a macro for all `Fq/p/p12/6...Repr` structs. All of them impl the same things over and over. So it's just boilerplate that we can save up.
- All of the things that are publicly exported but not docummented should do the exercise of: 1. Think if they need to be exported. 2. if so, add docs. Otherwise change visibility.

davidnevadoc commented on src/pluto_eris/fields/fp2.rs:
```diff
@@ -0,0 +1,771 @@
+    }
+
+    pub fn frobenius_map(&mut self, power: usize) {
+        //TODO Replace with constant time version if needed
```
Oh, that is quite counter-intuitive! Why is constant-time faster?
The previous version of this function was constant-time in the value of  `self` as well as `power`. Now, it is not constant in the value of `power`.
The current algorithm is:
```
if power %2 !=0:
   self.cojungate() // Conjugate is just negating the second coordinate
```
The previous algorithm was:
``` 
FROB_CONSTS = [ -1. 1]
self.c1.mul_assign( FROB_CONSTS[power%2])
```

CPerezz replied:
Ye, my brain turned off probably here. Of course the more constant usually the less optimal..  No strong feelings. But seeing the difference, I think we can go with the `const` one as seems will not change much. 
In any case, your call. No strong feelings here.

davidnevadoc commented on src/pluto_eris/curve.rs:
```diff
@@ -0,0 +1,283 @@
+    c1: Fp::ONE,
+};
+
+impl group::cofactor::CofactorGroup for G1 {
+    type Subgroup = G1;
+
+    fn clear_cofactor(&self) -> Self {
+        *self
+    }
+
+    fn into_subgroup(self) -> CtOption<Self::Subgroup> {
+        CtOption::new(self, 1.into())
+    }
+
+    fn is_torsion_free(&self) -> Choice {
+        1.into()
+    }
+}
```
Essentially yes, we need this trait for the `new_curve_impl!` macro. However, I think it is only used in the implementation of `random` to be able to clear the cofactor and return a valid point.

davidnevadoc commented on src/pluto_eris/curve.rs:
```diff
@@ -0,0 +1,283 @@
+    }
+
+    fn into_subgroup(self) -> CtOption<Self::Subgroup> {
+        unimplemented!()
```
This is inherited from the `bn246` implementation. I believe the reason for having this unimplemented is that we are not using the trait as it is intended.
We are implementing `CofactorGroup` for `G2`.  The associated type for `Subgroup` should not be `G2` but a different struct. 
In relation to that, the result of `clear_cofactor` should be `Self::Subgroup` not `Self`, it just happens that this is ok for most curves, since they have cofactor 1.

I will open an issue detailing this problem and other changes that also affect the implementation of `bn256`. In this way we can address them there as well.

davidnevadoc commented on src/pluto_eris/engine.rs:
```diff
@@ -0,0 +1,942 @@
+impl Group for Gt {
+    type Scalar = Fq;
+
+    fn random(_: impl RngCore) -> Self {
+        unimplemented!();
+    }
```
Again, this was inherited from `bn256`. The main reason is that this is never used, yet the trait `Group` is necessary for the implementation of the pairing.
I will add this to the issue I mentioned earlier to solve it in both curves.

davidnevadoc commented on src/pluto_eris/engine.rs:
```diff
@@ -0,0 +1,942 @@
+    }
+}
+
+pub fn multi_miller_loop(terms: &[(&G1Affine, &G2Prepared)]) -> Gt {
```
Good point! I have cleaned up the code a bit and implemented this directly as part of the `MultiMillerLoop` trait.

davidnevadoc commented on src/pluto_eris/engine.rs:
```diff
@@ -0,0 +1,942 @@
+    Gt(f)
+}
+
+pub fn pairing(g1: &G1Affine, g2: &G2Affine) -> Gt {
```
Done! ( Same case as `multi_miller_loop`)
Again, this should be fixed in `bn256` as well.

davidnevadoc commented on src/pluto_eris/fields/fp.rs:
```diff
@@ -0,0 +1,552 @@
+    /// Computes the multiplicative inverse of this element,
+    /// failing if the element is zero.
+    fn invert(&self) -> CtOption<Self> {
+        let tmp = self.pow([
```
Done. Also, we should switch to the BY method now that is merged.

davidnevadoc commented on src/pluto_eris/fields/fp12.rs:
```diff
@@ -0,0 +1,660 @@
+        self.double()
+    }
+
+    fn sqrt(&self) -> CtOption<Self> {
+        unimplemented!()
+    }
+
+    fn sqrt_ratio(_num: &Self, _div: &Self) -> (Choice, Self) {
+        unimplemented!()
+    }
```
The square root is only used when we want to find the y-coordinate of some point of a curve, so in general, it is not needed for fields over which we haven't defined any curve.

davidnevadoc replied:
Added as comment in the code!

davidnevadoc commented on src/pluto_eris/fields/fp12.rs:
```diff
@@ -0,0 +1,660 @@
+    },
+];
+
+#[cfg(test)]
+use rand::SeedableRng;
+#[cfg(test)]
+use rand_xorshift::XorShiftRng;
+
+#[test]
+fn test_fp12_mul_by_014() {
+    let mut rng = XorShiftRng::from_seed([
+        0x59, 0x62, 0xbe, 0x5d, 0x76, 0x3d, 0x31, 0x8d, 0x17, 0xdb, 0x37, 0x32, 0x54, 0x06, 0xbc,
+        0xe5,
+    ]);
+
+    for _ in 0..1000 {
+        let c0 = Fp2::random(&mut rng);
+        let c1 = Fp2::random(&mut rng);
```
I'm thinking of having a common test suite so I'll address this in https://github.com/privacy-scaling-explorations/halo2curves/issues/76.

davidnevadoc commented on src/pluto_eris/fields/fp2.rs:
```diff
@@ -0,0 +1,771 @@
+        res
+    }
+
+    // TODO: This is a naive method using 4 multiplications
```
When the non-residue used is `i` with `i^2 = -1` the algorithm can be improved to use 3 muls instead of 4. However, I think it's not possible to make such optimization in this case.
I'll look into it and remove the comment / apply the optimization if possible.

CPerezz replied:
Can you link the issue created for this and mark resolved?

davidnevadoc commented on src/pluto_eris/fields/fp2.rs:
```diff
@@ -0,0 +1,771 @@
+
+    // TODO: This is a naive method using 4 multiplications
+    pub fn mul_assign(&mut self, other: &Self) {
+        // r0 = s0 * s0 + U_SQUARE * s1 * o1
+        // r1 = s0 * o1 - s1 * o0
```
It's an explanation of the code below.

davidnevadoc commented on src/pluto_eris/fields/fp2.rs:
```diff
@@ -0,0 +1,771 @@
+        self.c1 = t1 + t2
+    }
+
+    // TODO: This is a naive method using 3 multiplications
+    pub fn square_assign(&mut self) {
+        // r0 = s0^2 + U_SQUARE * s1^2
+        // r1 = 2* s0s1
```
Same as the previous response.

davidnevadoc commented on src/pluto_eris/fields/fp6.rs:
```diff
@@ -0,0 +1,787 @@
+//     c1: Fp::from_raw([0x01, 0, 0, 0, 0, 0, 0]),
+// };
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Default)]
+pub struct Fp6 {
```
I don't think it is very useful to comment these fields. The description of the sturct `Fp6`:
`/// The `Fp6` element c0 + c1 * v + c2 * v3`
already provides all the information needed.

davidnevadoc commented on src/pluto_eris/fields/fq.rs:
```diff
@@ -0,0 +1,556 @@
+        end_timer!(start);
+    }
+
+    #[test]
+    fn bench_fp_from_u16() {
+        let repeat = 10000000;
+        let mut rng = ark_std::test_rng();
+        let base = (0..repeat).map(|_| (rng.next_u32() % (1 << 16)) as u64);
+
+        let timer = start_timer!(|| format!("generate {} Bn256 scalar field elements", repeat));
+        let _res: Vec<_> = base.map(Fq::from).collect();
+
+        end_timer!(timer);
+    }
```
Yes, this is a leftover. I will remove it and work on having common benchmarks for all fields and curves.

davidnevadoc commented on src/pluto_eris/curve.rs:
```diff
@@ -0,0 +1,283 @@
+}
+
+impl G1 {
+    const SVDW_Z: Fp = Fp::ONE;
```
Added

davidnevadoc commented on src/pluto_eris/curve.rs:
```diff
@@ -0,0 +1,283 @@
+}
+
+impl Eris {
+    const SVDW_Z: Fq = Fq::ONE;
```
Added

davidnevadoc commented on src/pluto_eris/engine.rs:
```diff
@@ -0,0 +1,942 @@
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, -1, 0, 1,
+];
+// the value of (57/(u + 3))^((p - 1)/2) where u^2 + 5 = 0 in Fp2
+pub const XI_TO_P_MINUS_1_OVER_2: Fp2 = Fp2 {
```
Removed `pub` and added docs.

davidnevadoc commented on src/pluto_eris/fields/fq.rs:
```diff
@@ -0,0 +1,556 @@
+    //             0xaaaaaaaaaaaaaaaa
+    //         ])
+    //     );
+    // }
```
New test vectors added, surprisingly they pass with the asm version as well! (see #100)

davidnevadoc commented on src/pluto_eris/fields/fq.rs:
```diff
@@ -0,0 +1,556 @@
+#[derive(Clone, Copy, Debug)]
+pub struct FqRepr {
+    pub repr: [u8; SIZE],
+}
```
Docs added.

davidnevadoc commented on src/pluto_eris/engine.rs:
```diff
@@ -0,0 +1,942 @@
+// Adaptation of Algorithm 1, https://eprint.iacr.org/2013/722.pdf
+// the parameter for the curve Pluto: u = -0x4000000000001000008780000000
+pub const NEG_PLUTO_U: u128 = 0x4000000000001000008780000000;
+pub const PLUTO_U_IS_NEGATIVE: bool = true;
```
Good catch! Removed.

davidnevadoc commented on src/pluto_eris/engine.rs:
```diff
@@ -0,0 +1,942 @@
+    /// Returns the group identity, which is $1$.
+    pub fn identity() -> Gt {
+        Gt(Fp12::ONE)
+    }
```
Added, should be modified in `bn256` as well.

davidnevadoc commented on src/pluto_eris/engine.rs:
```diff
@@ -0,0 +1,942 @@
+}
+
+impl G2Prepared {
+    pub fn is_zero(&self) -> bool {
```
Added.

davidnevadoc commented on src/pluto_eris/engine.rs:
```diff
@@ -0,0 +1,942 @@
+        self.infinity
+    }
+
+    pub fn from_affine(q: G2Affine) -> Self {
```
Added.

davidnevadoc commented on src/pluto_eris/engine.rs:
```diff
@@ -0,0 +1,942 @@
+            };
+        }
+
+        fn doubling_step(r: &mut G2) -> (Fp2, Fp2, Fp2) {
```
Done

davidnevadoc commented on src/pluto_eris/engine.rs:
```diff
@@ -0,0 +1,942 @@
+        for &mut (p, ref mut coeffs) in &mut pairs {
+            ell(&mut f, coeffs.next().unwrap(), p);
+        }
+        if x != 0 {
```
Fixed

davidnevadoc commented on src/pluto_eris/engine.rs:
```diff
@@ -0,0 +1,942 @@
+            coeffs.push(doubling_step(&mut r));
+
+            let x = NEG_SIX_U_PLUS_2_NAF[i];
+            match x {
```
Good point, I removed the variable to make the code a bit clearer.
However, this part of the code is still really messy. I will refactor it (and `bn256`'s) to make it more readable.

davidnevadoc commented on src/pluto_eris/fields/fp.rs:
```diff
@@ -0,0 +1,552 @@
+}
+
+#[derive(Clone, Copy, Debug)]
+pub struct FpRepr {
```
Done

davidnevadoc commented on src/pluto_eris/fields/fp.rs:
```diff
@@ -0,0 +1,552 @@
+    //             0xaaaaaaaaaaaaaaaa
+    //         ])
+    //     );
+    // }
```
New test vectors added!

davidnevadoc commented on src/pluto_eris/fields/fp12.rs:
```diff
@@ -0,0 +1,660 @@
+    }
+}
+
+// non_residue^((modulus^i-1)/6) for i=0,...,11
```
Added as docs and clarified the comment a bit.

davidnevadoc commented on src/pluto_eris/fields/fp2.rs:
```diff
@@ -0,0 +1,771 @@
+
+/// -ALPHA is a quadratic non-residue in Fp. Fp2 = Fp[X]/(X^2 + ALPHA)
+/// We introduce the variable u such that u^2 = -ALPHA
+// const ALPHA: u64 = 5;
```
This is a leftover, the const `U_SQUARE` is used instead.

davidnevadoc commented on src/pluto_eris/fields/fp2.rs:
```diff
@@ -0,0 +1,771 @@
+impl_sum_prod!(Fp2);
+
+const SIZE: usize = 112;
+const COORD_SIZE: usize = 56;
```
Added.

davidnevadoc commented on src/pluto_eris/fields/fp2.rs:
```diff
@@ -0,0 +1,771 @@
+    };
+}
+
+#[cfg(test)]
+use rand::SeedableRng;
+#[cfg(test)]
+use rand_xorshift::XorShiftRng;
+
+#[test]
+fn test_ser() {
```
Will be addressed in #76

davidnevadoc commented on src/pluto_eris/fields/fq.rs:
```diff
@@ -0,0 +1,556 @@
+}
+
+#[cfg(test)]
+mod test {
```
I think this would be too verbose, I find this clear enough:
`pluto_eris::fields::fq::test::`

davidnevadoc commented on src/pluto_eris/engine.rs:
```diff
@@ -0,0 +1,942 @@
+
+        coeffs.push(doubling_step(&mut r));
+
+        match NEG_SIX_U_PLUS_2_NAF[NEG_SIX_U_PLUS_2_NAF.len() - 2] {
```
I agree. Will refactor this code as part of a bn engine macro in a separate PR.

CPerezz replied:
Can you link the issue filled for this here?

davidnevadoc replied:
Will be addressed in https://github.com/privacy-scaling-explorations/halo2curves/issues/105

davidnevadoc commented on src/pluto_eris/engine.rs:
```diff
@@ -0,0 +1,942 @@
+            -1 => {
+                coeffs.push(addition_step(&mut r, &negq));
+            }
+            _ => (),
```
I think the best thing is to solve this in the bn pairing refactor.

CPerezz replied:
Can you link the issue and mark resolved?

davidnevadoc replied:
Will be addressed in https://github.com/privacy-scaling-explorations/halo2curves/issues/105

davidnevadoc commented on src/pluto_eris/fields/fp.rs:
```diff
@@ -0,0 +1,552 @@
+}
+
+impl FpRepr {
+    pub fn iter(&self) -> Iter<'_, u8> {
```
Done

davidnevadoc commented on src/pluto_eris/fields/fp6.rs:
```diff
@@ -0,0 +1,787 @@
+
+/// -BETA is a cubic non-residue in Fp2. Fp6 = Fp2[X]/(X^3 + BETA)
+/// We introduce the variable v such that v^3 = -BETA
+// BETA = - 57/(z+3)
```
Added as doc

davidnevadoc commented on src/pluto_eris/fields/fq.rs:
```diff
@@ -0,0 +1,556 @@
+//     0xa803ca76883922dc,
+//     0x0130e0000d7ee0e4,
+//     0x2400000000002400,
+// ]);
```
Removed

davidnevadoc commented on src/pluto_eris/fields/fp.rs:
```diff
@@ -0,0 +1,552 @@
+#[cfg_attr(feature = "derive_serde", derive(Serialize, Deserialize))]
+pub struct Fp(pub(crate) [u64; 7]);
+
+const SIZE: usize = 56;
```
Done

davidnevadoc commented:
I have addressed the comments relevant to this PR. However there are still some unresolved that I think should be fixed in a different PR. Mainly:
 1. **Tests and benchmarks**
      Fixing these consistently would involved modifying other curves which is outside of the scope of this PR. ( Just added the test vectors specific for this curves). This should be fixed in #76 
  2. **Documentation and unimplemented**
     There are some unimplemented and undocumented functions. These are unimplemented or undocumented in `bn256`.
     I plan to fix both issues in a single PR to avoid duplicating work. I also would like to introduce a common macro for the BN pairing and do some refactoring to improve the code quality in #105

davidnevadoc commented on src/pluto_eris/fields/fp6.rs:
```diff
@@ -0,0 +1,787 @@
+    },
+];
+
+#[cfg(test)]
+use rand::SeedableRng;
+#[cfg(test)]
+use rand_xorshift::XorShiftRng;
+
+#[test]
+fn test_fp6_mul_nonresidue() {
+    let mut rng = XorShiftRng::from_seed([
+        0x59, 0x62, 0xbe, 0x5d, 0x76, 0x3d, 0x31, 0x8d, 0x17, 0xdb, 0x37, 0x32, 0x54, 0x06, 0xbc,
+        0xe5,
+    ]);
+
```
Will be addressed in https://github.com/privacy-scaling-explorations/halo2curves/issues/76

davidnevadoc commented on src/pluto_eris/fields/fp2.rs:
```diff
@@ -0,0 +1,771 @@
+impl_binops_multiplicative!(Fp2, Fp2);
+impl_sum_prod!(Fp2);
+
+const SIZE: usize = 112;
```
Added, although GH didn't pick up the change.

davidnevadoc commented on src/pluto_eris/fields/fp2.rs:
```diff
@@ -0,0 +1,771 @@
+    }
+}
+#[derive(Clone, Copy, Debug)]
+pub struct Fp2Bytes([u8; SIZE]);
```
Added, although GH didn't pick it up.

davidnevadoc commented on src/pluto_eris/engine.rs:
```diff
@@ -0,0 +1,942 @@
+    }
+
+    // Final steps of the line function on prepared coefficients
+    fn ell(f: &mut Fp12, coeffs: &(Fp2, Fp2, Fp2), p: &G1Affine) {
```
I think this could be a good improvement!
I have been reading a bit about autovectorisation but it is unclear to me what are the conditions that need to be met for it to apply. 

This will be a good extra improvement for the refactor in #76.

CPerezz replied:
It's usually hard to tell as depends only on the compiler. But I know at least in the past, Iterators were guaranteed to take profit of it. Not for loops where it wasn't always the case though.

In any case, makes sense to address later!

davidnevadoc commented on src/pluto_eris/fields/fq.rs:
```diff
@@ -0,0 +1,556 @@
+#[cfg_attr(feature = "derive_serde", derive(Serialize, Deserialize))]
+pub struct Fq(pub(crate) [u64; 7]);
+
+const SIZE: usize = 56;
```
Done, although GH didn't pick it up

davidnevadoc commented on src/pluto_eris/engine.rs:
```diff
@@ -0,0 +1,942 @@
+                -1 => {
+                    coeffs.push(addition_step(&mut r, &negq));
+                }
+                _ => continue,
```
Will be addressed in #105

CPerezz reviewed (approved):
We are inconsistent between our comments.

We have the hex representation of only some values for the constants that are defined in this PR.

I think we should either have them all or none. But it's quite confusing to only have "some of them".

We can address in a follow-up PR if you want. But this has to be addressed at some point.

Other than that, LGTM! 

Good work with this!!

CPerezz commented:
cc: @daira this might be of your interest!
